File: frontend/src/utils/frameQuality.ts
// Shared frame-quality deviation utility.
//
// Grades a single frame metric against a learned MAD-based baseline and turns it
// into a signed robust z-score, a color band, and a combined 0-100 quality score.
// The group-key format ("telescope|camera|filter") matches the backend exactly so
// frontend and backend index the same baseline maps.

export type QualityBand = "better" | "neutral" | "watch" | "reject";

export interface MetricBaseline {
  median: number | null;
  mad: number | null;
  n: number;
}

export type GroupBaseline = Record<string, MetricBaseline>;
export type Baselines = Record<string, GroupBaseline>;

export const MIN_GROUP = 8;
const EPS = 1e-9;

export function groupKey(f: {
  telescope?: string | null;
  camera?: string | null;
  filter_used?: string | null;
}): string {
  return `${f.telescope ?? ""}|${f.camera ?? ""}|${f.filter_used ?? ""}`;
}

// MAD-based robust z-score. higherIsBetter flips the sign so that "worse than
// baseline" is always positive regardless of metric polarity.
// Returns null (treated as neutral) when the baseline is too sparse, uniform
// (mad === 0), or the value/baseline is missing.
export function madZ(
  value: number | null | undefined,
  b: MetricBaseline | undefined,
  higherIsBetter = false,
): number | null {
  if (value == null || !b || b.median == null || b.mad == null || b.n < MIN_GROUP || b.mad === 0) return null;
  const z = (value - b.median) / Math.max(b.mad, EPS);
  return higherIsBetter ? -z : z;
}

export function bandForZ(z: number | null): QualityBand {
  if (z == null) return "neutral";
  if (z <= -1.0) return "better";
  if (z < 1.5) return "neutral";
  if (z < 3.0) return "watch";
  return "reject";
}

// Combined 0-100 score from up to three axes (signal/sharpness/roundness).
// Weights 0.5 / 0.25 / 0.25; when an axis is null it is dropped and the
// remaining weights are renormalized to sum to 1. All-null -> null (neutral row).
export function combinedScore(
  zSignal: number | null,
  zSharp: number | null,
  zRound: number | null,
): number | null {
  const axes = [
    [zSignal, 0.5],
    [zSharp, 0.25],
    [zRound, 0.25],
  ].filter(([z]) => z != null) as [number, number][];
  if (axes.length === 0) return null;
  const wsum = axes.reduce((s, [, w]) => s + w, 0);
  const zc = axes.reduce((s, [z, w]) => s + z * w, 0) / wsum;
  const clamped = Math.max(-3, Math.min(3, zc));
  return 50 - 16.7 * clamped;
}

export function bandToCellClass(band: QualityBand): string {
  switch (band) {
    case "better": return "text-theme-success";
    case "watch":  return "text-theme-warning";
    case "reject": return "text-theme-error";
    default:       return "text-theme-text-primary";
  }
}

export function scoreToRowClass(score: number | null): string {
  if (score == null) return "";
  if (score >= 60) return "bg-theme-success/10";
  if (score >= 45) return "";
  if (score >= 30) return "bg-theme-warning/15";
  return "bg-theme-error/20";
}

// Worked examples (sanity checks for the math above):
//
// 1. Higher-is-worse metric (HFR): value 2.8, baseline median 2.1, mad 0.37.
//    z = (2.8 - 2.1) / 0.37 = 0.7 / 0.37 ≈ 1.89  ->  bandForZ -> "watch".
//
// 2. Higher-is-better metric (detected_stars): value 80, baseline median 100,
//    mad 10, higherIsBetter = true.
//    raw z = (80 - 100) / 10 = -2; sign-flipped -> +2  ->  bandForZ -> "watch".
//    (Fewer stars than baseline is worse, so the flipped z is positive.)

File: backend/app/services/frame_quality.py
"""MAD-based frame-quality baseline math.

Pure functions with no DB access. They compute, for a collection of already
loaded light frames grouped by optical train + filter, a robust baseline
(median + median-absolute-deviation) per graded metric. The frontend turns a
per-frame metric and the matching baseline into a signed z-score and color band.

No new DB columns and no DATA_VERSION bump: baselines are derived on the fly
from existing per-frame data.
"""

import statistics

MIN_GROUP = 8
EPS = 1e-9

# Deviation at which a frame is called an outlier, in RAW MAD units (there is no
# 1.4826 consistency scaling anywhere in this codebase, so these are not sigma).
# The value matches the frontend `bandForZ` "reject" band in
# frontend/src/utils/frameQuality.ts, so a session insight and the per-frame cell
# colours rendered beside it cannot contradict each other.
Z_REJECT = 3.0

# Metrics for which a baseline (median + MAD) is computed per group.
METRICS = [
    "median_hfr",
    "fwhm",
    "eccentricity",
    "detected_stars",
    "adu_median",
    "guiding_rms_arcsec",
]


def median(values):
    """Median of values, ignoring None. Returns None for an empty input.

    For an even count the average of the two middle values is returned, matching
    statistics.median.
    """
    nums = [v for v in values if v is not None]
    if not nums:
        return None
    return statistics.median(nums)


def mad(values):
    """Median absolute deviation from the median, ignoring None.

    Returns None for an empty input and 0.0 for a uniform group.
    """
    nums = [v for v in values if v is not None]
    if not nums:
        return None
    m = statistics.median(nums)
    return statistics.median([abs(x - m) for x in nums])


def _get(frame, key):
    """Read a field from a frame that may be a dict or an attribute-bearing object."""
    if isinstance(frame, dict):
        return frame.get(key)
    return getattr(frame, key, None)


def group_key(frame) -> str:
    """Build the train+filter group key matching the frontend `groupKey(frame)`.

    Format: ``"{telescope}|{camera}|{filter_used}"`` with null components
    rendered as an empty string.
    """
    tel = _get(frame, "telescope") or ""
    cam = _get(frame, "camera") or ""
    filt = _get(frame, "filter_used") or ""
    return f"{tel}|{cam}|{filt}"


def group_baselines(frames):
    """Compute per-group, per-metric ``{median, mad, n}`` baselines.

    `frames` may be dicts or attribute-bearing objects. Groups by `group_key`;
    for each metric, n is the count of non-null values for that metric within
    the group.
    """
    groups: dict[str, list] = {}
    for frame in frames:
        groups.setdefault(group_key(frame), []).append(frame)

    out: dict[str, dict] = {}
    for key, group_frames in groups.items():
        metric_stats: dict[str, dict] = {}
        for metric in METRICS:
            vals = [_get(f, metric) for f in group_frames]
            non_null = [v for v in vals if v is not None]
            metric_stats[metric] = {
                "median": median(non_null),
                "mad": mad(non_null),
                "n": len(non_null),
            }
        out[key] = metric_stats
    return out


def mad_z(value, baseline, higher_is_better=False):
    """Signed robust z-score of `value` against a ``{median, mad, n}`` baseline.

    Deliberately mirrors the frontend `madZ` (frontend/src/utils/frameQuality.ts)
    including its null semantics, so backend insight text and frontend cell
    colours are computed from the same number. Returns None when:

    - the value is missing,
    - the baseline is missing or has no median/mad,
    - the group is too sparse (n < MIN_GROUP), or
    - the group is uniform (mad == 0), which would divide by ~nothing.

    A None z must never produce an insight: staying silent beats a wrong claim.

    `higher_is_better` flips the sign so "worse than baseline" is always
    positive. Units are raw MAD, not sigma.
    """
    if value is None or not baseline:
        return None
    med = baseline.get("median")
    dev = baseline.get("mad")
    n = baseline.get("n") or 0
    if med is None or dev is None or n < MIN_GROUP or dev == 0:
        return None
    z = (value - med) / max(dev, EPS)
    return -z if higher_is_better else z


def count_outliers(frames, baselines, metric, *, z_threshold=Z_REJECT, higher_is_better=False):
    """Count frames whose `metric` sits at or beyond `z_threshold` MAD units on
    the bad side of the baseline for that frame's OWN train+filter group.

    Returns ``(outliers, graded)``. `graded` is the number of frames that had a
    usable baseline, letting callers distinguish "nothing was bad" from "nothing
    could be judged" and stay silent in the second case.
    """
    outliers = 0
    graded = 0
    for frame in frames or []:
        group = baselines.get(group_key(frame)) if baselines else None
        z = mad_z(_get(frame, metric), (group or {}).get(metric), higher_is_better)
        if z is None:
            continue
        graded += 1
        if z >= z_threshold:
            outliers += 1
    return outliers, graded
